internals/overlord/logstate/loki.go

```diff
@@ -0,0 +1,157 @@
+// Copyright (c) 2023 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package logstate
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/canonical/pebble/internals/plan"
+	"github.com/canonical/pebble/internals/servicelog"
+)
+
+const (
+	maxLokiRequestSize = 10
+	lokiRequestTimeout = 10 * time.Second
+)
+
+type lokiClient struct {
+	target    *plan.LogTarget
+	remoteURL string
+	// buffered entries are "sharded" by service name
+	entries    map[string][]lokiEntry
+	numEntries int
+
+	httpClient http.Client
+}
+
+func newLokiClient(target *plan.LogTarget) logClient {
+	return &lokiClient{
+		target:     target,
+		remoteURL:  target.Location,
+		entries:    map[string][]lokiEntry{},
+		httpClient: http.Client{Timeout: lokiRequestTimeout},
+	}
+}
+
+func (c *lokiClient) Write(ctx context.Context, entry servicelog.Entry) error {
+	c.entries[entry.Service] = append(c.entries[entry.Service], asLokiEntry(entry))
+	c.numEntries++
+	if c.numEntries >= maxLokiRequestSize {
+		return c.Flush(ctx)
+	}
+	return nil
+}
+
+func asLokiEntry(entry servicelog.Entry) lokiEntry {
+	return lokiEntry{
+		strconv.FormatInt(entry.Time.UnixNano(), 10),
+		strings.TrimSuffix(entry.Message, "\n"),
+	}
+}
+
+func (c *lokiClient) Flush(ctx context.Context) error {
+	if c.numEntries == 0 {
+		return nil // no-op
+	}
+	defer c.emptyBuffer()
+
+	req := c.buildRequest()
+	jsonReq, err := json.Marshal(req)
+	if err != nil {
+		return fmt.Errorf("encoding request to JSON: %v", err)
+	}
+
+	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, c.remoteURL, bytes.NewReader(jsonReq))
+	if err != nil {
+		return fmt.Errorf("creating HTTP request: %v", err)
+	}
+	httpReq.Header.Set("Content-Type", "application/json")
+
+	resp, err := c.httpClient.Do(httpReq)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		// Drain request body to allow connection reuse
+		// see https://pkg.go.dev/net/http#Response.Body
+		_, _ = io.Copy(io.Discard, io.LimitReader(resp.Body, 1024*1024))
+		_ = resp.Body.Close()
+	}()
+
+	// Check response status code to see if it was successful
+	if code := resp.StatusCode; code < 200 || code >= 300 {
+		// Request to Loki failed
+		b, err := io.ReadAll(io.LimitReader(resp.Body, 1024))
```

Comment:
Is this limit (1024) to protect us against error responses with too much detail?

Reply:
essentially. If we're reading more than 1KB from the response, we're probably just spamming the user's terminal and not really giving any more useful info than what is described in the first 1KB.